### PR TITLE
add card models

### LINKED ISSUE #5: Intitial cards model and data population of mongoDB.
**EDIT:** Add some basic Card.java `@Document`class, that is extensible enough to serve as the base class for all our flash cards documents - i.e. the instances that make up the collection 
-> consider abstract or sealed class - but not exaggerated complexity. Derived classes could be SimpleQAndACard, MultipleChoiceCard, and may well be records)
*--- end of edit to data model --- added, since it is probably needed for the population task below ---*

Add some initial data to populate the mongoDB:

- a sample user (to the user collection)
- some initial card data - as an example collection `db.example`.

The data are well-suited for us now in early development phase, as they can be used for implementing display / modification views. But they can also stay in permanently and serve to new users to flashcards as public card sample collection to get familiar.

Incrementally, the example collection should consist of different card types as simple Q&A, multiple choice, etc.. to illustrate the options a user has.

Several alternatives - increasingly more effort - on how to implement the task (all alternative should only insert, if the example collection does not yet exist)

1. **Insert manually per script**. 
Use init script to insert data  - like 5 to 10 Json records for a start (comparable to `spring.sql.init.data-locations=classpath:initData.sql`)
2. **Insert manually via CommandLineRunner**. 
same as 1 only Commandlinerunner
3. **use quiz data from the net**.  
find and download some suitable (maybe a little larger) quiz data set and transform it into our cards example collection by use of a CommandLineRunner.
4. **scrape website and generate cards**
e.g. use the problems of some Hyperskill topic group (e.g. Spring)

### LINKED ISSUE #13: Create flashcard models
Extracted from Issue #5:
Add some basic Card.java @Documentclass, that is extensible enough to serve as the base class for all our flash cards documents - i.e. the instances that make up the collection
-> consider abstract or sealed class - but not exaggerated complexity. Derived classes could be SimpleQAndACard, MultipleChoiceCard, and may well be records)

### PR DESCRIPTION
**Pull request by alfabravo2013**
A draft of card models that requires more use case analysis.

### CODE REVIEW

File: server/src/main/java/org/hyperskill/community/flashcards/config/MongoConfiguration.java [hunk 1]
```diff
@@ -1,16 +1,24 @@
...
     public MongoTemplate mongoTemplate() {
         return new MongoTemplate(MongoClients.create(), "cards");
     }
 
+    @Bean
+    @Qualifier("example")
+    public MongoTemplate exampleMongoTemplate() {
+        return new MongoTemplate(MongoClients.create(), "example");
+    }
 }
```
Comment (wisskirchenj) [inline server/src/main/java/org/hyperskill/community/flashcards/config/MongoConfiguration.java:24]: This bean creates the example database.
It's not was I originally intendend - as I thought more of creating a collection named "example" within the cards database.

But we can discuss this anyway:

Problem: This ansatz to structure cards into several databases (example would be the first one) might collide with the multi-user and access concept of issue #5.
If user a (database A) shares one collection to user b (database B), then when we query B's collections, we need to query over both databases... So in general, we need to query all databases to find the cards accessible to a soecific user.

File: server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+        log.info("Dropping any existing collections in 'example' database...");
+        var collections = mongoTemplate.getDb().listCollectionNames();
+
+        log.info("Inserting sample data to 'example' database...");
+        cards.forEach(card -> {
+            var category = card.getCategory();
+            mongoTemplate.insert(card, category);
+        });
+    }
```
Comment (wisskirchenj) [inline server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java]: Ah you thought of a full example **database**? I gathered, we'd put all cards into the _collection_ named "example". (see comment at MongoConfiguration)

File: server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+@ToString
+public abstract class Card {
+    @Id
+    protected String id;
+    protected String author;
+    protected boolean isPublic;
+    protected String title;
+    protected String category;
+    protected String question;
+    @CreatedDate
```
Comment (wisskirchenj) [inline server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java:27]: do we need to store the category - or isn't this more the collection, the document is attahced to - i.e. do we need this back-reference?

File: server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ * Basic class for all flashcards, contains fields shared among all subclasses
+ */
+@Getter
+@Setter
+@Document
+@ToString
+public abstract class Card {
+    @Id
```
Comment (wisskirchenj) [inline server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java:22]: Wouldn't card be a perfect use case for a sealed class ... permits MultipleChoiceQuiz, SingleChoiceQuiz, QuestionAndAnswerCard

Then we could make perfect use of Java 21 pattern matching also.

File: server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                            return qac;
+                        } else if (card instanceof SingleChoiceQuiz scc) {
+                            return scc;
+                        } else if (card instanceof MultipleChoiceQuiz mcc) {
+                            return mcc;
+                        } else {
+                            throw new IllegalStateException("Unknown card class " + card.getClass().getName());
+                        }
+                    })
+                    .map(Objects::toString)
```
Comment (wisskirchenj) [inline server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java]: what's the map with instanceof for here? You can apply toString without it.
and the forEach could read .forEach(log::debug) to use the logger (or info if we really want to log each card every time :-))

File: server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                        .isPublic(true)
+                        .question("Which of the elements has the lowest atomic weight?")
+                        .options(List.of("Au", "Fe", "Li", "Pt"))
+                        .correctOption(2)
+                        .build()
+        );
+
+        log.info("Dropping any existing collections in 'example' database...");
+        var collections = mongoTemplate.getDb().listCollectionNames();
+
```
Comment (wisskirchenj) [inline server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java:72]: line 80 can be removed - unused

File: server/src/main/java/org/hyperskill/community/flashcards/card/model/MultipleChoiceQuiz.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+ * having options with checkboxes
+ */
+@Getter
+@Setter
+@ToString(callSuper = true)
+@TypeAlias("MultipleChoiceQuiz")
+public class MultipleChoiceQuiz extends Card {
+    private List<String> options;
+    private List<Integer> correctOptions;
+
```
Comment (wisskirchenj) [inline server/src/main/java/org/hyperskill/community/flashcards/card/model/MultipleChoiceQuiz.java:24]: is it possible to use Java records here - or does this come with cons?
question applies to all three subtypes of course.

File: server/src/main/java/org/hyperskill/community/flashcards/config/MongoConfiguration.java (hunk 1 above)
Comment (alfabravo2013) [reply server/src/main/java/org/hyperskill/community/flashcards/config/MongoConfiguration.java:24]: Let's keep it for now, I'll remove it when the card and category models are finalized.

File: server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java (hunk 2 above)
Comment (alfabravo2013) [reply server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java]: See above :arrow_up:

File: server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java (hunk 3 above)
Comment (alfabravo2013) [reply server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java:27]: Agree, removed this backing field,

File: server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java (hunk 4 above)
Comment (alfabravo2013) [reply server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java:22]: Agreed, changed to a sealed class with three subclasses.

File: server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java (hunk 5 above)
Comment (alfabravo2013) [reply server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java]: For demo purposes only, to proof that we can actually use polymorfic collections without any issues.

File: server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java (hunk 6 above)
Comment (alfabravo2013) [reply server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java:72]: Forgot to call dropCollection, fixed.

File: server/src/main/java/org/hyperskill/community/flashcards/card/model/MultipleChoiceQuiz.java (hunk 7 above)
Comment (alfabravo2013) [reply server/src/main/java/org/hyperskill/community/flashcards/card/model/MultipleChoiceQuiz.java:24]: Records cannot extend classes and we have quite a few shared properties. Having a sealed interface is less beneficial in my opinion.

File: server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java [hunk 8]
```diff
@@ -101,18 +102,12 @@ public void run(String... args) {
...
-                        }
+            getAllCards(collectionName).stream()
+                    .map(card -> switch (card) {
+                        case SingleChoiceQuiz scq -> scq;
+                        case MultipleChoiceQuiz mcq -> mcq;
+                        case QuestionAndAnswerCard qac -> qac;
                     })
-                    .map(Objects::toString)
                     .forEach(System.out::println);
         }
```
Comment (wisskirchenj) [inline server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java]: still: whatis the purpose of the map() here?
Why not just:
getAllCards(collectionName).foreach(c -> log.debug("" + c); // or log.info if you prefer for now

File: server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java (hunk 5 above)
Comment (wisskirchenj) [reply server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java]: Ah i see. Ok it's gone now anyway

File: server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java (hunk 8 above)
Comment (alfabravo2013) [reply server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java]: To prove that cards are successfully cast to their class. It doesn't make sense to have tests at this stage, so it's a sort of substitution for them. Will remove the unnecessary queries and log entries in the final commit.

File: server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java [hunk 9]
```diff
@@ -21,7 +20,7 @@
  */
 @Component
 @Slf4j
-public class ExampleDataInitializer implements CommandLineRunner {
+public class ExampleDataInitializer {
     private final MongoTemplate mongoTemplate;
```
Comment (wisskirchenj) [inline server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java:25]: you removed the CLR? so it is not called on startup. intentionally?

File: server/src/main/java/org/hyperskill/community/flashcards/category/model/CategoryAssess.java [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+
+/**
+ * Represents a set of permissions for a given user
+ *
+ * @param username   the name of the user
+ * @param permission a string that defines the set of permissions.
+ *                   The possible permissions are: <b>r</b> for read, <b>w</b> for write and <b>d</b> for delete
+ */
+public record CategoryAssess(String username, String permission) {
+}
```
Comment (wisskirchenj) [inline server/src/main/java/org/hyperskill/community/flashcards/category/model/CategoryAssess.java]: typo - should be CategoryAccess

File: server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java [hunk 11]
```diff
@@ -82,44 +76,13 @@ public void init() {
...
-            getAllCards(collectionName).stream()
-                    .map(card -> switch (card) {
-                        case SingleChoiceQuiz scq -> scq;
-                        case MultipleChoiceQuiz mcq -> mcq;
-                        case QuestionAndAnswerCard qac -> qac;
-                    })
-                    .forEach(System.out::println);
-        }
+        var categoryAccess = new CategoryAssess("test@test.com", "rwd");
+        mongoTemplate.insert(new Category(null, "math", Set.of(categoryAccess)));
```
Comment (wisskirchenj) [inline server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java:75]: variable name categoryAccess

File: server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java [hunk 12]
```diff
@@ -31,9 +30,8 @@ public abstract sealed class Card permits QuestionAndAnswerCard, SingleChoiceQui
 
     public Card() { }
```
Comment (wisskirchenj) [inline server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java]: what is the public constructor for? does mongo need it?

**Review by wisskirchenj — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java (hunk 9 above)
Comment (wisskirchenj) [reply server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java:25]: done via @PostConstruct - so fine

Comment (alfabravo2013) [reply server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java:25]: `@PostConstruct` does the job.

File: server/src/main/java/org/hyperskill/community/flashcards/category/model/CategoryAssess.java (hunk 10 above)
Comment (alfabravo2013) [reply server/src/main/java/org/hyperskill/community/flashcards/category/model/CategoryAssess.java]: fixed

File: server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java (hunk 12 above)
Comment (alfabravo2013) [reply server/src/main/java/org/hyperskill/community/flashcards/card/model/Card.java]: Removed unused constructors, they are not required.

File: server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java (hunk 11 above)
Comment (alfabravo2013) [reply server/src/main/java/org/hyperskill/community/flashcards/card/ExampleDataInitializer.java:75]: fixed

**Review by wisskirchenj — APPROVED** [review]

**Pull request merged**